pkgs/development/python-modules/pydot_ng/default.nix

```
@@ -16,13 +17,17 @@ buildPythonPackage rec {
     sha256 = "8c8073b97aa7030c28118961e2c6c92f046e4cb57aeba7df87146f7baa6530c5";
   };
 
-  buildInputs = [ pytest unittest2 ];
+  patches = [ ./0001-graphviz_path.patch ];
+
+  postPatch =
+    ''
+    sed -i 's#NIX_GRAPHVIZ_PATH#"${pkgs.graphviz}/bin"#g' pydot_ng/__init__.py;
```

There's a nice utility to do this (although i would just apply the other changes, this is more FYSA)

```suggestion
    substituteInPlace pydot_ng/__init__.py \
      -- replace "NIX_GRAPHVIZ_PATH" "${pkgs.graphviz}/bin"
```

---

If we just modify the `PATH` then the tests will pass, but surely at module import time there's no guarantee that `dot` will be found (or that it's the same `dot` we ran the tests against). Apologies if that's what you're already saying, I'm a little confused by your comment.

I agree depending on the exact text is a bit ugly.

---

So, i took another stab at it, just adding a link to the dot executable after the wheel installation:

<details> <summary> New expression </summary>

```
$ cat pkgs/development/python-modules/pydot_ng/default.nix
{ stdenv
, buildPythonPackage
, fetchPypi
, pyparsing
, pytest
, unittest2
, pkgs
, mock
}:

buildPythonPackage rec {
  pname = "pydot_ng";
  version = "2.0.0";

  src = fetchPypi {
    inherit pname version;
    sha256 = "8c8073b97aa7030c28118961e2c6c92f046e4cb57aeba7df87146f7baa6530c5";
  };

  # executable is needed so it can be shelled commands
  postInstall = ''
    mkdir  $out/bin
    ln -s ${pkgs.graphviz}/bin/dot $out/bin/dot
  '';

  checkInputs = [ pytest unittest2 mock ];
  propagatedBuildInputs = [ pkgs.graphviz pyparsing ];

  checkPhase = "PATH=$PATH:$out/bin py.test test";

  meta = with stdenv.lib; {
    homepage = "https://pypi.python.org/pypi/pydot-ng";
    description = "Python 3-compatible update of pydot, a Python interface to Graphviz's Dot";
    license = licenses.mit;
    maintainers = [ maintainers.bcdarwin ];
  };

}
```

</details>

But it does work!

```
[23:30:06] jon@jon-desktop /home/jon/projects/nixpkgs (fix/644461-pydot_ng)
$ nix-shell -p 'with import ./. {}; pythonPackages.pydot_ng' --pure -p "with pythonPackages; [ pytest mock nose]"

[nix-shell:/home/jon/projects/nixpkgs]$ cd ../pydot-ng/
ChangeLog  LICENSE  MANIFEST.in  README.rst  appveyor.yml  docs  pydot_ng  requirements-conda.txt  setup.cfg  setup.py  test  test-requirements.txt  tox.ini

[nix-shell:/home/jon/projects/pydot-ng]$ pytest .
===================================================================================================================== test session starts =====================================================================================================================
platform linux2 -- Python 2.7.16, pytest-4.6.3, py-1.7.0, pluggy-0.12.0
rootdir: /home/jon/projects/pydot-ng
collected 330 items

test/test_graph_from_edges.py ..........................................................                                                                                                                                                                [ 17%]
test/test_pydot.py ......................................................................                                                                                                                                                               [ 38%]
test/test_rendering_dot_files.py ..........................................................................................................................................................................................................             [100%]

================================================================================================================= 330 passed in 57.28 seconds =================================================================================================================
```

Someone might be able to think of a better way of ensuring the dot executable is on the path, but I can't think of one :)

---

Patching is the preferred solution.

---

It makes the tests pass, but I don't think it works in general. If I try your expression:
```
$ cat > pkgs/development/python-modules/pydot_ng/default.nix
...
<EOF>
$ nix-build -E 'with import ./. {}; python37.withPackages (p: [p.pydot_ng])'
$ ./result/bin/python -c "import pydot_ng; print (pydot_ng.find_graphviz())"
None
```
It can't find `dot` because even though it is inside `result/bin`, that's not in `$PATH`. So the tests will pass, however the module will not work in practice unless you happen to have `dot` in your path anyway, I think. For comparison if you manually set `$PATH` when you import then it works:
```
PATH=result/bin ./result/bin/python -c "import pydot_ng; print (pydot_ng.find_graphviz())"
{'dot': 'result/bin/dot', 'twopi': '', 'neato': '', 'circo': '', 'fdp': '', 'sfdp': ''}
```

---

Ah, i see, since i used nix-shell, it was giving me an environment with it on path. nix-build would not.

Still wish there was a better way to propagate an executable. :(

---

Aside: did you want substituteInPlace? I think it's nicer than plain sed.

Better way: betas me - I'm new to all this Nix stuff. If the python module took a `--graphviz-path=` type option then we wouldn't have a problem.

---

Yea :) . I find substituteInPlace a lot easier to read. [More shell commands if you're interested](https://nixos.org/nixpkgs/manual/#ssec-stdenv-functions)

And I've only been actively contributing for a few months, still learning new things all the time, like in this thread :)